src/mesh.cpp

```diff
@@ -1,80 +1,123 @@
-void readMainFile(Mesh &mesh, const std::string &filename)
+void readMainFile(Mesh &mesh, const std::string &filename, const std::string &dataname, const int &dim)
 {
+
   if (!fs::is_regular_file(filename)) {
     throw std::invalid_argument{"The mesh file does not exist: " + filename};
   }
-  std::ifstream mainFile{filename};
-  std::string   line;
-  while (std::getline(mainFile, line)) {
-    double             x, y, z, val;
-    std::istringstream iss(line);
-    iss >> x >> y >> z >> val; // split up by whitespace
-    std::array<double, 3> vertexPos{x, y, z};
-    mesh.positions.push_back(vertexPos);
-    mesh.data.push_back(val);
-  }
-}
 
-// Reads the connectivity file containing the triangle and edge information
-void readConnFile(Mesh &mesh, const std::string &filename)
-{
-  if (!fs::is_regular_file(filename)) {
-    throw std::invalid_argument{"The mesh connectivity file does not exist: " + filename};
+  vtkSmartPointer<vtkGenericDataObjectReader> reader =
+      vtkSmartPointer<vtkGenericDataObjectReader>::New();
+  reader->SetFileName(filename.c_str());
+  reader->SetReadAllScalars(true);
+  reader->SetReadAllVectors(true);
+  reader->ReadAllFieldsOn();
+  reader->Update();
+
+  // Get Points
+  vtkSmartPointer<vtkPoints> Points    = reader->GetUnstructuredGridOutput()->GetPoints();
+  vtkIdType                  NumPoints = reader->GetUnstructuredGridOutput()->GetNumberOfPoints();
+  for (vtkIdType point = 0; point < NumPoints; point++) {
+    std::array<double, 3> vertexPosArr;
+    Points->GetPoint(point, vertexPosArr.data());
+    mesh.positions.push_back(vertexPosArr);
   }
-  std::ifstream connFile{filename};
-  std::string   line;
-  while (std::getline(connFile, line)) {
-    std::vector<std::string> parts;
-    boost::split(parts, line, [](char c) { return c == ' '; });
-    std::vector<size_t> indices(parts.size());
-    std::transform(parts.begin(), parts.end(), indices.begin(), [](const std::string &s) -> size_t { return std::stol(s); });
-
-    if (indices.size() == 3) {
-      std::array<size_t, 3> elem{indices[0], indices[1], indices[2]};
-      mesh.triangles.push_back(elem);
-    } else if (indices.size() == 2) {
-      std::array<size_t, 2> elem{indices[0], indices[1]};
-      mesh.edges.push_back(elem);
-    } else {
-      throw std::runtime_error{std::string{"Invalid entry in connectivitiy file \""}.append(line).append("\"")};
+  // Get Point Data
+  vtkSmartPointer<vtkPointData> PD = reader->GetUnstructuredGridOutput()->GetPointData();
+  // Check it has data array
+  if (PD->HasArray(dataname.c_str()) == 1) {
+    // Get Data and Add to Mesh
+    vtkSmartPointer<vtkDataArray> ArrayData = PD->GetArray(dataname.c_str());
+    int                           NumComp   = ArrayData->GetNumberOfComponents();
+
+    if (NumComp != dim) {
+      throw std::runtime_error("Dimensions of data provided and simulation does not match!.");
     }
+    // Reserve enough space for data
+    mesh.data.reserve(NumPoints * dim);
```

Comment:
The code below only works if `mesh.data` is empty.
We should either `assert()` this, or `clean()` the data vector.